compiler/rustc_mir/src/transform/instcombine.rs

```diff
@@ -87,6 +93,58 @@ impl OptimizationFinder<'b, 'tcx> {
         OptimizationFinder { body, tcx, optimizations: OptimizationList::default() }
     }
 
+    fn find_deref_of_address(&mut self, rvalue: &Rvalue<'tcx>, location: Location) -> Option<()> {
+        // look for the sequence
+        // _2 = &_1;
+        // ...
+        // _5 = (*_2)
+        //
+        // which we can replace the last statement with `_5 = _1` to avoid the load of _2
+        if let Rvalue::Use(op) = rvalue {
+            let local_being_derefed = match op.place()?.as_ref() {
+                PlaceRef { local, projection: [ProjectionElem::Deref] } => Some(local),
+                _ => None,
+            }?;
+
+            let stmt_index = location.statement_index;
+            // look behind for statement that assigns the local from a address of operator
+            // 6 is chosen as a heuristic determined by seeing the number of times the optimization kicked in compiling rust std
+            let lower_index = stmt_index.saturating_sub(6);
+            let statements_to_look_in = self.body.basic_blocks()[location.block].statements
+                [lower_index..stmt_index]
+                .iter()
+                .rev();
+            for stmt in statements_to_look_in {
+                match &stmt.kind {
+                    rustc_middle::mir::StatementKind::Assign(box (l, r))
```

Comment:
you also need to bail out on `LlvmAssembly`. Please exhaustively match the `StatementKind`s so we catch all uses.

Reply:
I added exhaustive matches, and now bail on ASM